shared/AppInsightsCore/src/JavaScriptSDK/BaseTelemetryPlugin.ts

```diff
@@ -0,0 +1,129 @@
+     * can optionally use this to access the current core instance or define / pass additional information
+     * to later plugins (vs appending items to the telemetry item)
+     */
+    public processNext: (env: ITelemetryItem, itemCtx: ITelemetryItemContext) => void;
+
+    /**
+     * Set next extension for telemetry processing
+     */
+    public setNextPlugin: (next: ITelemetryPlugin) => void;
+
+    /**
+     * Returns the current logger that can be used to log issues, if no logger is currently
+     * assigned a new default one will be created and returned.
+     */
+    public log: (itemCtx?:ITelemetryItemContext) => IDiagnosticLogger;
+
+    /**
+     * Returns whether the plugin has been initialized
+     */
+    public isInitialized: () => boolean;
+
+    public identifier: string;
+    public version?: string;
+
+    /**
+     * Holds the core instance that was used during initialization
+     */
+    public core: IAppInsightsCore;              
+
+    priority: number;
+
+    /**
+     * Helper to return the current ITelemetryItemContext, if the passed argument exists this just
+     * returns that value (helps with minification for callers), otherwise it will return the configured
+     * context or a temporary one.
+     * @param overrideItemCtx - [Optional] Override the default context
+     */
+    protected _getTelCtx: (overrideItemCtx?:ITelemetryItemContext) => ITelemetryItemContext;
+
+    /**
+     * Accessor to allow setting the default ITelemetryContext used during initialization and unit testing
+     */
+    protected _setTelCtx: (telemetryCtx: ITelemetryItemContext) => void;
+
+    /**
+     * Internal helper to allow setting of the internal initialized setting for inherited instances and unit testing
+     */
+    protected setInitialized: (isInitialized: boolean) => void;
+
+    constructor() {
+        let _self = this;
+        let _isinitialized = false;
+        let _defTelContext: ITelemetryItemContext = null;
+
+        _self.core = null;
+
+        _self.log = (itemCtx:ITelemetryItemContext): IDiagnosticLogger => {
+            return _self._getTelCtx(itemCtx).log();
+        }
+
+        _self.isInitialized = () => {
+            return _isinitialized;
+        }
+
+        _self.setInitialized = (isInitialized: boolean):void => {
+            _isinitialized = isInitialized;
+        }
+
+        _self.setNextPlugin = (next: ITelemetryPlugin) => {
+            // Does nothing as plugin sub-classes of this base class should be using 
+            // processNext() function for execution or the passed ITelemetryItemContext.getNext().getPlugin()
+            // if direct access to the next plugin is required. Note getNext() may return null.
+        };
+
+        _self.processNext = (env: ITelemetryItem, itemCtx: ITelemetryItemContext) => {
+            _self._getTelCtx(itemCtx).processNext(env);
+        };
+
+        _self._getTelCtx = (overrideCtx:ITelemetryItemContext = null) => {
```

Comment:
This needs work :-(
For proper backward compatibility I need to record and process the instance level nextPlugin chain when no telemetryItemContext is available.